FLINK-2168 Add HBaseTableSource

ramkrish86 opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

@fhueske 
Trying to create the first version of this PR. I have made the necessary changes to support HBaseTableSource by creating a HBaseTableInputFormat but lot of code is duplicated with TableInputFormat. I have not unified them for now.
I tried compiling this code in my linux box but the @Override that I have added in HBaseTableSource overriding the BatchTableSource API shows as compilation issues but my IntelliJ IDE is fine and does not complain.
Pls provide your valuable feed back so that I can rebase the next PR with suitable fixes. 
Thanks.

ramkrish86 commented:
And one more thing is that other than the BasicTypeInfo what other types should we support. I was not sure on that so added a TODO there.

tonycox commented:
You need to recompile `TableSource` trait manually and implement `DefinedFieldNames` in `HBaseTableSource`

fhueske commented:
@tonycox is right. Please rebase your PR to the current master branch. The `TableSource` interface was recently modified.

ramkrish86 commented:
I see. Am not sure how I missed that because my IDE I thought was already updated with latest code. Will check it.

tonycox commented:
@ramkrish86 @fhueske what do you think about to throw `Tuple` (`T extends Tuple`)  out of `org.apache.flink.addons.hbase.TableInputFormat` and implement this abstract class in your `HBaseTableSourceInputFormat` ?

ramkrish86 commented:
Thanks @tonycox .Yes I am fine with it. I can try. Anyway I t hink there is more to do from my side. Now am not sure how to register the table with the valid family name and column name. It is only registering with the table and it is not resolving '.select("f1:q1, f1:q2, f1:q3");'.
I am able to run the test case only now and so finding these. I will wait for comments and then go on with the updation of the PR.

tonycox commented:
I think it can be solved in different issue to provide a new api in `table.api` for selecting from HBase

tonycox commented:
And could you extend with `StreamTableSource` also ?

ramkrish86 commented:
So am not sure if we can add specific API in table.api for hbase? Is that allowed?

tonycox commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -0,0 +1,322 @@
+					values[i] = new Date(Bytes.toLong(value));
+				}
+			} else {
+				// TODO for other types??
```
Arrays, Java SQL Date/Time/Timestamp, CompositeType

ramkrish86 replied:
I can see that in the TableSourceITCase that uses CsvInputFormat - the metadata is also passed as the input and the meta data is registered with the table. Now in cases like hbase - how should that be done?

ramkrish86 replied:
{code}
testHBaseTableSource(org.apache.flink.addons.hbase.example.HBaseTableSourceITCase)  Time elapsed: 2.297 sec  <<< ERROR!
org.apache.flink.table.api.ValidationException: Cannot resolve [q1] given input [f0, f1, f2].
        at org.apache.flink.addons.hbase.example.HBaseTableSourceITCase.testHBaseTableSource(HBaseTableSourceITCase.java:113)
{code}
The error says field input [f0, f1, f2]. Not sure how it got picked up.

tonycox replied:
I think you need to implement `DefinedFieldNames` trait in `HBaseTableSource`

wuchong replied:
No need to implement `DefinedFieldNames`. The [f0, f1, f2] is the default field names of RowTypeInfo, you should set the custom field names in RowTypeInfo. Using `RowTypeInfo(TypeInformation<?>[] types, String[] fieldNames)` constructor.

ramkrish86 replied:
HBaseTableSchema - So can we add the startKey and endKey also into this then?

wuchong replied:
I think it's fine to add start/endKey into HBaseTableSchema.

tonycox commented:
We need discuss that

ramkrish86 commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,117 @@
+		//2nd qual is String
+		put.addColumn(F_1, Q_2, Bytes.toBytes("strvalue2"));
+		// 3rd qual is long
+		put.addColumn(F_1, Q_2, Bytes.toBytes(19993l));
```
Ya.  Good catch. Will change it. I was more focussed on the impl because if you see the test case itself was not asserting any. I need to add all those stuff.

tonycox commented:
As Jark Wu said in [jira](https://issues.apache.org/jira/browse/FLINK-5554) 

> I think the HBaseTableSource should return a composite type (with column family and qualifier), and we can get columns by composite type accessing.

ramkrish86 commented:
Thanks for the ping here @tonycox .

wuchong commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -0,0 +1,322 @@
+
+	@Override
+	public TypeInformation<Row> getProducedType() {
+		return new RowTypeInfo(this.fieldTypeInfos);
```
Using `RowTypeInfo(TypeInformation<?>[] types, String[] fieldNames)` constructor to set the custom field names.

ramkrish86 replied:
Ok. Got it.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java:
```diff
@@ -0,0 +1,75 @@
+							TypeInformation<?>[] colTypes) {
+		this.conf = conf;
+		this.tableName = Preconditions.checkNotNull(tableName, "Table  name");
+		this.rowKey = Preconditions.checkNotNull(rowKey, "Rowkey");
```
Yes. That is true but do we always want full table scan? Actually in HBase it is better we specify start and end key. So how do we specify that? I have not used this rowKey now but I thought it is better to be used?

wuchong replied:
Make sense. We can support the start and end rowkey.

ramkrish86 replied:
`new RowTypeInfo(
	new TypeInformation[]{
		new RowTypeInfo(
			new TypeInformation[]{
				BasicTypeInfo.STRING_TYPE_INFO, BasicTypeInfo.INT_TYPE_INFO},
			new String[]{"name", "age"})
	},
	new String[]{"person"}
);`
I did the above change for the HBaseTableSource's return type and then made a simple query
`		tableEnv.registerTableSource("test", hbaseTable);
		Table result = tableEnv
			.sql("SELECT f1.q1, f1.q2, f1.q3 FROM test");`
It throws sql validation error and considers f1 as table name and says tableName 'f1' not found.

tonycox replied:
Try to use `.sql("SELECT test.f1.q1, test.f1.q2t")` or table api instead. I think there is a problem with nested Rows while scaning

wuchong replied:
Hi @ramkrish86 , the RowTypeInfo contains field types and also field names (default f1~fn) of the table. By using `RowTypeInfo(TypeInformation[] types, String[] fieldNames)` constructor, we can customize the field names of the table. So in your case, you can query columns by `.sql(SELECT person.name, person.age FROM test)`.

tonycox replied:
` f0~fn-1 `

ramkrish86 replied:
I tried this customization but still if I pass something like f1.q1 it was throwing a validation error which was fine after I used it as suggested by @tonycox .

fhueske commented:
Hi @ramkrish86, @tonycox, and @wuchong,

sorry for joining the discussion a bit late. I haven't looked at the code yet, but I think the discussion is going into the right direction. 

I had a look at [how Apache Drill provides access to HBase tables](https://drill.apache.org/docs/querying-hbase/). Drill also uses a nested schema of `[rowkey, colfamily1[col1, col2, ...], colfamiliy2[col1, col2, ...] ...]` so basically the same as we are discussing here.

Regarding the field types: The serialization is not under our control, so should also offer to just return the raw bytes (as Drill does). If users have custom data types or serialization logic they can use a user defined scalar function to extract the value. I don't know what's the standard serialization format for primitives with HBase (or if there is one at all). 

Regarding restricting the scan with rowkeys. @tonycox's PR for [filterable TableSources](https://github.com/apache/flink/pull/3166) can be used to set the scan range. This would be much better than "hardcoding" the scan ranges in the TableSource.

Best, Fabian

wuchong commented:
Hi @fhueske , 

Regarding to the field type serialization, I think maybe we can provide default deserialization for basic types (int,long,String...) if users do use the `Bytes.toBytes(...)` to serialize the basic types. If not, users can ask this field to return raw bytes in this way:  `htableSchame.add("column_family", "qualifier", byte[].class)` and then use an user defined scalar function to deserialize the value.

regarding to the rowkeys, I agree with you. It would be great if we can set scan range by WHERE clause. But FLINK-3849 (FilterableTableSource) is still a pending PR, I would suggest to break this issue into two. 1. add HBaseTableSource, provide access to HBase tables and support nested schema. 2. extend HBaseTableSource to support FilterableTableSource.

fhueske commented:
You are right @wuchong, we should break it down into two issues. 

I also agree about the serialization. We should offer defaults for primitives (`byte`, `short`, `int`, `long`, `boolean`, `float`, `double`) and a set of common character encodings (UTF-8, ASCII, etc.) for `String`. Everything else can be initially handled as `byte[]`, IMO.

ramkrish86 commented:
Thanks for all the inputs here. I have been trying to make my existing code work with the composite RowTypeInfo. Once that is done I will try to introduce the HBaseTableSchema.
Also I would like to work on FLINK-3849 (FilterableTableSource) after this first version of HBaseTableSource is accepted.

ramkrish86 commented:
Good news is that with the help of this Composite RowType and modifying my code accordingly and debugging things I could get the basic thing to work. Now I will work on stitching things together and submitting a PR with updated changes.

wuchong commented:
Sounds good ! Looking forward that !

ramkrish86 commented:
@fhueske , @tonycox , @wuchong 
I have updated the PR based on all the feedbacks here. Now you could see that we now support CompoisteRowType and we are able to specify multiple column families along with the qualifier names.
We are able to retrieve the result by doing a full scan.
This is not efficient and we need to specify start and end rows. I think that can be done after FilterableTableSource is done.
I have added test cases that shows single column family and double column family. 
For now if the TypeInformation is not known we use plain byte[] type only. That happens at the validation state itself. But one main concern from my side is how to present the 'NULL' means we specify a column with a type but there is no data for that column. For now I have handled by returning the Int, Float, Long - Min_values. But that may not be right I believe. Feedback and suggestions welcome.

wuchong commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -19,99 +19,113 @@
+		List<Object> values = new ArrayList<Object>();
+		int i = 0;
+		Map<String, List<Pair>> familyMap = schema.getFamilyMap();
+		Row[] rows = new Row[familyMap.size()];
```
Better to declare `rows` as `Object[]` to avoid confusing whether `rows` is a varargs or non-varargs.

ramkrish86 replied:
Will check this too. Thanks for all the nice comments @wuchong .

wuchong commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -0,0 +1,160 @@
+				if(value != null) {
+					values.add(schema.deserialize(value, pair.getSecond()));
+				} else {
+					values.add(schema.deserializeNull(pair.getSecond()));
```
Do we really need this method to indicate `null` using the special default value? Why not set null `values.add(null)` directly ? 

`Row` supports nullable field, but Tuple doesn't.

wuchong replied:
>But one main concern from my side is how to present the 'NULL' means we specify a column with a type but there is no data for that column. For now I have handled by returning the Int, Float, Long - Min_values. But that may not be right I believe. Feedback and suggestions welcome.

I think we can return `null` directly.

wuchong replied:
>But one main concern from my side is how to present the 'NULL' means we specify a column with a type but there is no data for that column. For now I have handled by returning the Int, Float, Long - Min_values. But that may not be right I believe. Feedback and suggestions welcome.

I think we can return `null` directly.

ramkrish86 replied:
Ya. I dint know this. Your question answers my doubt and the shaky method that I added can be removed.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,135 @@
+		Integer.class, Short.class, Float.class, Long.class, String.class, Byte.class, Boolean.class, Double.class, BigInteger.class, BigDecimal.class, Date.class
+	};
+	private static byte[] EMPTY_BYTE_ARRAY = new byte[0];
+	public void addColumns(String family, String qualifier, TypeInformation<?> type) {
```
Ok. Let me check that. So if we pass Class there we could wrap it with the corresponding TypeInformation for our internal usage?

wuchong replied:
Yes. We can use `TypeExtractor.getForClass(Class)` to extract the corresponding TypeInformation for internal use.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,135 @@
+		}
+		if(!found) {
+			// by default it will be byte[] type only
+			type = BasicArrayTypeInfo.BYTE_ARRAY_TYPE_INFO;
```
Ok. Got it.

ramkrish86 replied:
Ok I will add byte[].class into the CLASS_TYPES. And anything other than the ones in CLASS_TYPES we can throw an exception.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,135 @@
+			// by default it will be byte[] type only
+			type = BasicArrayTypeInfo.BYTE_ARRAY_TYPE_INFO;
+		}
+		list.add(new Pair(qualifier, type));
```
Ok

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java:
```diff
@@ -22,54 +22,63 @@
-		this.colTypes = Preconditions.checkNotNull(colTypes, "Field types");
+		this.schema = Preconditions.checkNotNull(schema, "Schema");
+		Map<String, List<Pair>> familyMap = schema.getFamilyMap();
+		famNames = familyMap.keySet().toArray(new String[familyMap.size()]);
```
Fine. Will do it.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java:
```diff
@@ -22,54 +22,63 @@
-	public ProjectableTableSource<Row> projectFields(int[] fields) {
-		String[] newColNames = new String[fields.length];
-		TypeInformation<?>[] newColTypes =  new TypeInformation<?>[fields.length];
+		List<String> qualNames = new ArrayList<String>();
```
I will look into this.

ex00 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,135 @@
+
+	public Object deserializeNull(TypeInformation<?> typeInfo) {
+		// TODO : this may need better handling.
+		if(typeInfo.getTypeClass() ==  Integer.class) {
```
Why don't try use here ``org.apache.flink.table.codegen.CodeGenUtils#primitiveDefaultValue``  for basic types and ?

wuchong replied:
I don't think so. `CodeGenUtils#primitiveDefaultValue` is used to give a default value to primitives. Flink code generation use primitives to improve performance. But the field of `Row` is not primitive which can be `null`.

ex00 replied:
`Row` type isn't basic type. I think may add check `typeInfo.isBasicType()` in this method like as in `#deserialize`.  and for not primitive types could return other default values.

ramkrish86 replied:
But returning default values may not be right. I just did that here as I was not aware of that NULL handling. When a column does not have any value then it is better we return NULL rather than any default value. @ex00 - What do you think?

ramkrish86 commented:
Updated the code with the comments and have pushed again. I think I have addressed all the comments here. Feedback/comments welcome. I also found that it is better to use the TableInputSplit to specify the start and end row so that the scan is anyway restricted to the given range.

ramkrish86 commented:
@fhueske , @tonycox , @wuchong - FYI.

tonycox commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java:
```diff
@@ -0,0 +1,65 @@
+ * For eg, cf1:q1 - where cf1 is the familyName and q1 is the qualifier name
+ */
+// TODO : Implement ProjectableTableSource?
+public class HBaseTableSource implements BatchTableSource<Row> {
```
will be `StreamTableSource` okay for HBase source?
I think it can be implemented same as batch `getDataSet`

ramkrish86 replied:
Am not sure.. For now I think we will implement BatchTableSource only and later implement StreamTableSource? Is there any significant design expectation for a source to be StreamTableSource?

tonycox commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,248 @@
+	public static final byte[] F_2 = Bytes.toBytes("f2");
+	public static final byte[] Q_1 = Bytes.toBytes("q1");
+	public static final byte[] Q_2 = Bytes.toBytes("q2");
+	public static final byte[] Q_3 = Bytes.toBytes("q3");
```
why public?

ramkrish86 replied:
ya . My bad. Will remove.

tonycox commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,248 @@
+		//2nd qual is String
+		put.addColumn(F_1, Q_2, Bytes.toBytes("strvalue2"));
+		// 3rd qual is long
+		put.addColumn(F_1, Q_3, Bytes.toBytes(19993l));
```
could you change it to upper L?
low l sometimes looks like 1

ramkrish86 replied:
ok

tonycox commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,248 @@
+	}
+
+
+	static <T> void compareResult(List<T> result, String expected, boolean asTuples, boolean sort) {
```
big copypaste, i think there is a better way - decouple `TestBaseUtils` in two classes - cluster starter and result checker. I will start conversation in dev list

wuchong replied:
@tonycox is right. This is a copypaste. But the methods in `TestBaseUtils` are static, we do not need to decouple it. We can use it directly like following:

```java
TestBaseUtils.compareResultAsText(results, expected);
```

Of course, we should add the `flink-test-utils` dependency first.

```
               <dependency>
			<groupId>org.apache.flink</groupId>
			<artifactId>flink-test-utils_2.10</artifactId>
			<version>${project.version}</version>
			<scope>test</scope>
		</dependency>
```

ramkrish86 replied:
I did not want to make the pom change and hence did this way.

wuchong replied:
I think it's fine to add a test scope dependency. The `flink-test-utils_2.10` is used in many modules including `flink-connector-kafka`s.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,137 @@
+
+	// A Map with key as column family.
+	private final Map<String, List<Pair<String, TypeInformation<?>>>> familyMap =
+		new HashMap<String, List<Pair<String, TypeInformation<?>>>>();
```
Ok. In our other projects we used to qualify the generic on both the sides.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,137 @@
+		new HashMap<String, List<Pair<String, TypeInformation<?>>>>();
+
+	// Allowed types. This may change.
+	// TODO : Check if the Date type should be the one in java.util or the one in java.sql
```
Ok. Makes sense.

ramkrish86 commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,137 @@
+	 */
+	public void addColumn(String family, String qualifier, Class<?> clazz) {
+		Preconditions.checkNotNull(family, "family name");
+		Preconditions.checkNotNull(family, "qualifier name");
```
Good catch.

ramkrish86 commented:
Fixed all the minor comments given above. @tonycox , @wuchong , @fhueske .

wuchong commented on flink-connectors/flink-hbase/pom.xml:
```diff
@@ -236,6 +246,12 @@ under the License.
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-test-utils_2.10</artifactId>
+			<version>1.3-SNAPSHOT</version>
```
Please use `${project.version}` here,  do not hard code the version.

wuchong commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,140 @@
+		}
+		if (!found) {
+			// throw exception
+			throw new IllegalArgumentException("Unsupported class type found " + clazz);
```
I think we can improve the error message to suggest users to use `byte[].class` instead and deserialize the byte array using user-defined scalar function.

wuchong commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,140 @@
+		return qualifierNames;
+	}
+
+	public TypeInformation<?>[] getTypeInformation(String family) {
```
Rename to `getQualifierTypes` ?  This can keep consistent to `getQualifierNames`. The `getTypeInformation` is a little mislead IMO.

wuchong commented on flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java:
```diff
@@ -464,7 +464,7 @@ protected static File asFile(String path) {
 	}
 
-	private static <T> void compareResult(List<T> result, String expected, boolean asTuples, boolean sort) {
+	public static <T> void compareResult(List<T> result, String expected, boolean asTuples, boolean sort) {
```
Can you revert this ?

wuchong commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,196 @@
+		String expected = "100,strvalue,19991,null,null,null\n" +
+			"null,null,null,201,newvalue1,29992\n" +
+			"102,strvalue2,19993,null,null,null\n";
+		TestBaseUtils.compareResult(results, expected, false, false);
```
Please use `TestBaseUtils.compareResultAsText(results, expected);`.

ramkrish86 replied:
Actually this API was expecting things to be sorted. I did not want that to happen. Ok then will ensure that the assertion is done after sorting. Thanks for all the comments. Will update the PR shortly.

ramkrish86 commented:
@wuchong 
I think I have updated the last comments from you. Thank you for all your help/support here.

ramkrish86 commented:
> Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "RpcServer.reader=6,bindAddress=testing-docker-bb4f2e37-e79f-42a3-a9e9-4995e42c70ba,port=45919"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CHAIN DataSource (at getDataSet(HBaseTableSource.java:63) (org.apache.flink.addons.hbase.HBaseTableSourceInputFormat)) -> FlatMap (select: (f1.q1 AS q1, f1.q2 AS q2, f1.q3 AS q3)) (6/32)"
01/27/2017 05:57:12	DataSink (collect())(13/32) switched to DEPLOYING 
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "LeaseRenewer:travis@localhost:39289"
01/27/2017 05:57:14	DataSink (collect())(12/32) switched to SCHEDULED 
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CHAIN DataSource (at getDataSet(HBaseTableSource.java:63) (org.apache.flink.addons.hbase.HBaseTableSourceInputFormat)) -> FlatMap (select: (f1.q1 AS q1, f1.q2 AS q2, f1.q3 AS q3)) (10/32)"

Getting this error in the travis build. But in my linux box it seems to pass.

ramkrish86 commented:
@fhueske ,
I have fixed the comments as per @wuchong . And he has said +1 after fixing them. Would like to see the PR and see if it is fine with you too.

tonycox commented:
@ramkrish86 I think you need implement projectable interface in HBase source

ramkrish86 commented:
For ProjectableTableSource I think I need some clarity. Because currently is based on int[] representing the fields. So am not sure how to map them in terms of qualifiers under a family.

wuchong commented:
I think the OutOfMemoryError error has nothing to do with this PR.

tonycox commented:
`HBaseTableSchema` is initial information of table columns, so we can add more maps that generate after familyMap has set
```java
Map<String, List<Pair<String, TypeInformation<?>>>> familyMap;
Map<String, Integer> familyMapId;
Map<String, Integer> columnMapId;
```

ramkrish86 commented:
Am not sure how we will manage this. But one thing that can be done is that if invalid family name is added to the scan - HBase internally throws FamilyNotFoundException - so that we can track and report back. But am not sure how to track the schema. So every time he wants to scan two families - the user has to call HBaseTableSchema#addColumns() twice (with the required qualifiers). 
@tonycox 
Can you help me understand - when you say ' so we can add more maps that generate after familyMap has set'?

tonycox commented:
@ramkrish86 sorry for confusing, don't pay attention to new generating maps.
I think `PushProjectIntoBatchTableSourceScanRule` is not good enough for nested data types. but we can project at least family columns now. 
`HBaseTableSchema` keep `familyMap` and might keep a list with family names.
for example, while `addColumn`
```java
List<Pair<String, TypeInformation<?>>> list = this.familyMap.get(family);
    if (list == null) {
        familyNames.add(family);
        list = new ArrayList<>();
    }
```
so we have access to family column by index

tonycox commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,140 @@
+	// Allowed types. This may change.
+	// TODO : Check if the Date type should be the one in java.util or the one in java.sql
+	private static Class[] CLASS_TYPES = {
+		Integer.class, Short.class, Float.class, Long.class, String.class, Byte.class, Boolean.class, Double.class, BigInteger.class, BigDecimal.class, Date.class, Time.class, byte[].class
```
why is this static class?
what do you think to use an immutable list? for example `com.google.common.collect.ImmutableCollection`
```java
private static ImmutableCollection<Class<?>> CLASS_TYPES = ImmutableList.<Class<?>>of(...
```

ramkrish86 replied:
Ok. I can do. I have not used much of these Immutblecollection. Infact in our project we removed the dependency on this. Hence went on with  a default way.

tonycox commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,140 @@
+				break;
+			}
+		}
+		if (!found) {
```
and just `if (!CLASS_TYPES.contains(clazz))` here

tonycox commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java:
```diff
@@ -0,0 +1,140 @@
+public class HBaseTableSchema implements Serializable {
+
+	// A Map with key as column family.
+	private final Map<String, List<Pair<String, TypeInformation<?>>>> familyMap =
```
why List of Pairs?
maybe should use Map instead?
```java
Map<String, Map<String, TypeInformation<?>>> familyMap
```

ramkrish86 replied:
I just felt Map is not needed here. Anyway we have the Key as String. If you are so particular then I can change.

ramkrish86 replied:
> I think PushProjectIntoBatchTableSourceScanRule is not good enough for nested data types. but we can project at least family columns now.

So you mean we will project the column families. One thing to remember is that though we have two families - Eduction and Department. When we retrieve Department would come first and then Education. So what matters is how we retrieve the result.

ramkrish86 commented:
@tonycox 
I have addressed all your latest comments including making HBaseTableSource a ProjectableTableSource. 
@wuchong , @fhueske 
Are you guys fine with the latest updates. If so we can try closing this PR and further discussions of adding StreamingtableSource and supporting WHERE clauses can be done in subsequent PRs.

fhueske commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Creates a table source that helps to scan data from an hbase table
+ *
+ * Note : the colNames are specified along with a familyName and they are seperated by a ':'
```
I think this comment is no longer valid as columns are now nested.

ramkrish86 replied:
True. Will remove.

fhueske commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -0,0 +1,144 @@
+/**
+ * {@link InputFormat} subclass that wraps the access for HTables. Returns the result as {@link Row}
+ */
+public class HBaseTableSourceInputFormat extends TableInputFormat<Row> implements ResultTypeQueryable<Row> {
```
Rename to `HBaseRowInputFormat`?

ramkrish86 replied:
Here the Row - you mean the Type of generic that is passed to this class? Because the HBase table also has the concept of row which we are now not using it. So I feel the current name is better. What you think?

fhueske replied:
This class could also be used without the `TableSource` from a regular DataSet program. I think the name `HBaseRowInputFormat` would make that more clear and in addition indicate that it returns `Row` objects.

fhueske commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -0,0 +1,144 @@
+	@Override
+	protected Scan getScanner() {
+		Scan scan = new Scan();
+		for(String family : schema.getFamilyNames()) {
```
Is it guaranteed that `schema.getFamilyNames()` always returns the names in the same order?

ramkrish86 replied:
Actually in hbse every thing is lexographically sorted (with byte[]). But here since we represent using String the map that we use in HBaseTableSchema should be sorting this in the natural order. 
But the results that we retrieve will always be lexographically sorted and we try to retrieve the specific cell from the result using this

>  res.getValue(Bytes.toBytes(family), Bytes.toBytes(qualifier));

fhueske commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java:
```diff
@@ -0,0 +1,144 @@
+	protected Scan getScanner() {
+		Scan scan = new Scan();
+		for(String family : schema.getFamilyNames()) {
+			for(String qualifierName : schema.getQualifierNames(family)) {
```
Is the order of qualifier names guaranteed?

fhueske commented on flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/TableInputFormat.java:
```diff
@@ -41,7 +41,7 @@
  * {@link InputFormat} subclass that wraps the access for HTables.
  */
-public abstract class TableInputFormat<T extends Tuple> extends RichInputFormat<T, TableInputSplit> {
+public abstract class TableInputFormat<T> extends RichInputFormat<T, TableInputSplit> {
```
I'd rather rename this class to `AbstractTableInputFormat<T>` and add a new class `TableInputFormat<T extends Tuple>` to keep the API stable.

fhueske commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,196 @@
+		table.put(puts);
+		table.close();
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+		BatchTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env, 	new TableConfig());
```
remove spaces after `env,`

fhueske commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,196 @@
+		// add the mutations to the table
+		table.put(puts);
+		table.close();
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
```
add a new line and a comment that the preparation is done.

fhueske commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseTableSourceITCase.java:
```diff
@@ -0,0 +1,196 @@
+		// add the mutations to the table
+		table.put(puts);
+		table.close();
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
```
same comments as in test above

ramkrish86 commented:
To understand better

>  We could make flat schema an optional mode or implement it as a separate TableSource as well.
and this one

> This could be solved if we use a flat schema and encode the nesting as columnFamily$column

Are you talking about using seperators for it? May be am not getting your concern here. Ya I agree that nested schema is better API but if we go with flat schema then maintaining the family to qualifier relation may not be easy. As you said a seperate TableSource where we define such things would be better.
Regarding HBaseTableSchema I think that is better so that we could modify that class for better serialization and deserialization by adding more logic for different types of classes. Even when we go with flat schema I think this type of class would help us to maintain the logic of family to qualifier mapping?

ramkrish86 commented:
Updated the PR fixing the comments. The comments were simple but the adding AbstractTableInputFormat and moving the code back and forth makes this one a bigger change. But internally they are just refactorings. The test cases passes.

fhueske commented:
The points I raised effect the API and different people tend to have different opinions about APIs ;-).

I suggested to use a flat schema with that names columns `columnFamily$qualifier`, i.e., no nesting but composing the column name from `colFamily` and `qualifier` and separating them by `$`. Internally we can and should still use the `family`-`qualifier` but just map everything to a flat schema. The question is whether this would that make the `HBaseTableSource` harder to use. I don't think the column access (`family.qualifier` vs. `family$qualifier`) would be much harder, but acting with families that have lots of columns would be more cumbersome, because each column would be a top-level column and would need to be explicitly selected in a `SELECT` clause. In the nested case, all columns of a family are conveniently grouped together.

Regarding the `HBaseTableSchema`, we could also use it only internally and not expose it to the user. The `HBaseTableSource` would have a method `addColumn()` and forward the calls to its internal `HBaseSchema`.

ramkrish86 commented:
Just a general question
`def projectFields(fields: Array[Int]): ProjectableTableSource[T]`

Is it mandatory to only have int[] here. Can we have String[] that allows to specify specific names? May be even more generic could be to have a ColumnarTableProjectableSource which allows to specify family to column mapping some way? Ultimately its the table source that is going to do the mapping and create the projected table source. So that should be fine?

ramkrish86 commented:
> Regarding the HBaseTableSchema, we could also use it only internally and not expose it to the user. The HBaseTableSource would have a method addColumn() and forward the calls to its internal HBaseSchema.

Have done this. I initially thought to do things in the construction time itself. Now added an addColumn() in hbaseTableSource. HBaseSchema becomes totally package private and no access to users.

Regarding the flatSchema, generally in hbase only family is required and the qualifiers are just dynamic. But here for the sake of accessibility we expect the user to specify the column names. (trying to give a relational look). It is in case of projections is where we have some issues. In my opinion if we have a better API for Projection may be we could handle it better? The current nested way as you said is better in the sense that all columns of a family are grouped together.

tonycox commented:
@fhueske I preffer a better API with nesting. We can extend `ProjectableTableSource` with a method like this
```scala
def projectNestedFields(fields: Array[String]): ProjectableTableSource[T]
```
with default realisation.
override the method in HBase table source and while pushing projection down extract `fieldAccessor`s from `inputRex` and push them into as well.
Or create another `ProjectableNestedFieldsTableSource` with the same method.

ramkrish86 commented:
`		Table result = tableEnv
			.sql("SELECT test1.f1.q1, test1.f2.q2 FROM test1 where test1.f1.q1 < 103");`

I just tried this query and it works with or without ProjectableTableSource.  So just wanted to know when does the projection come into place. I thought without Projection things of this sort may not work.

tonycox commented:
@ramkrish86 ProjectableTableSource works in scan process. Without it TableScan is scanning all collumns and applying flatMap function on the whole data to project and filter. It's inefficient compared with pushing a projection and filtering right inside of scan proccess

ramkrish86 commented:
I think going through the PR for https://issues.apache.org/jira/browse/FLINK-3848 - I think we try to project only the required columns. Similarly we could do here also. So my and @tonycox 's suggestion of having a new way of ProjectableTableSource could help here.

ramkrish86 commented:
bq.ProjectableTableSource works in scan process. 
Ya got it. I was just trying to relate with this HBase thing and could find that we try to read all cols and then do a flatMap and then return the required cols alone. Just read that PR to understand better.

ramkrish86 commented:
But there are actually few more things as HBase can do if the scan queries are intelligent. Am not sure how to do them. We have filters that can be pushed to the server along with the scans. That can do some intelligent work on the server side itself. May be all that for later PRs.

wuchong commented:
Sorry for the late response. 

Regarding to the `HBaseTableSchema`, I agree with that to move the `addColumn(...)` method into `HBaseTableSource`. 

Regarding to the nested vs flat schema, I prefer the nested schema. It is more intuitive to use.
As for the nested schema doesn't support to push projections down, I think we should extend `ProjectableTableSource` to support push projections down to a composite type. We can keep the interface unchanged, i.e. `def projectFields(fields: Array[Int]): ProjectableTableSource[T]`. But the index of `fields` should be the flat index. We can use the flat field indexes to do projection pushdown even if it is a nested schema.

For example, a table source with schema `a: Int, b: Row<b1: String, b2: Int>, c: Boolean`, the flat indexes of `a, b.b1, b.b2, c` are `0, 1, 2, 3`. So a project `SELECT b.b1, c FROM T` will result a `fields` `Array(1,3)`.

What do you think ? 


For me the biggest drawback of a nested schema is the lacking support to push projections down.

fhueske commented:
Hi all,

thanks for the feedback. Let's stick to the nested schema then.

I think the best approach to support projections on nested fields is to implement a second interface (i.e., a trait without default implementation) called `NestedFieldsProjectableTableSource` as @tonycox suggested.
Adding a method with default implementation to `ProjectableTableSource` would not work, because this would turn this class into a Java abstract class while it is an interface now.
Using flat indicies is not a very nice solution either, IMO because it is not easy to parse.

For now I'd suggest to keep the scope of the PR as it is right now. A bit more Java documentation on `HBaseTableSource` to explain how it is used would be great.
We can implement the `NestedFieldsProjectableTableSource` and the changes to `HBaseTableSource` in a follow up issue.

What do you think?

ramkrish86 commented:
> For now I'd suggest to keep the scope of the PR as it is right now. A bit more Java documentation on HBaseTableSource to explain how it is used would be great.
We can implement the NestedFieldsProjectableTableSource and the changes to HBaseTableSource in a follow up issue.

+1 for this. I can add some more javadoc to it. BTW am trying to checkout these Projections and using the ProjectabletableSource. Will be back on it. Thanks to every one for all the comments and feedback.

ramkrish86 commented:
@fhueske - Please have a look at the javadoc.

wuchong commented:
The `NestedFieldsProjectableTableSource` makes sense to me. +1 to implement it in a separate JIRA.

fhueske commented:
Thanks for the update @ramkrish86 and your patience!
I skimmed over the PR and it looks good.
I'll do a final pass early next week before merging it.

Thanks, Fabian

ramkrish86 commented:
@fhueske - A gentle reminder !!!

fhueske commented:
Alright. The PR looks mostly good. I'll make a few refinements and will merge it then.

fhueske commented:
I observed an `OutOfMemoryError` in one of the Travis profiles when running the final tests. Will need to figure out what's the reason for that before I can merge the PR.

ramkrish86 commented:
> Will need to figure out what's the reason for that before I can merge the PR.

I tried running those tests again in my linux box and all went through without any error.

ramkrish86 commented:
`02/02/2017 06:19:20	DataSink (collect())(6/32) switched to SCHEDULED 
02/02/2017 06:19:20	DataSink (collect())(2/32) switched to SCHEDULED 
02/02/2017 06:19:20	DataSink (collect())(8/32) switched to SCHEDULED 
02/02/2017 06:19:20	DataSink (collect())(6/32) switched to DEPLOYING 
02/02/2017 06:19:20	DataSink (collect())(8/32) switched to DEPLOYING 
02/02/2017 06:19:20	DataSink (collect())(2/32) switched to DEPLOYING 
02/02/2017 06:19:25	DataSink (collect())(1/32) switched to SCHEDULED 
02/02/2017 06:19:25	DataSink (collect())(1/32) switched to DEPLOYING 
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "JvmPauseMonitor"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "hconnection-0x2247c79b-shared--pool20-t1"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "JvmPauseMonitor"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "flink-akka.actor.default-dispatcher-3"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "IPC Server handler 1 on 34267"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CHAIN DataSource (localhost:53456)"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CHAIN DataSource (localhost:53456)"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "CHAIN DataSource (localhost:53456)"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "org.apache.hadoop.hdfs.PeerCache@28360f4a"`

The same tests run cleanly in my linux box. Is there any place I can download the logs to see what could be the issue?

tonycox commented:
Try to run in flink folder (I ran it onto ubuntu 16.04)
```bash
export JAVA_HOME=/path/to/java-7-oracle/jre
mvn -B -Dhadoop.version=2.3.0 \
test -Dtest=org.apache.flink.addons.hbase.example.HBaseTableSourceITCase \
-DfailIfNoTests=false
```

ramkrish86 commented:
I tried even that. The test runs fine for me. No OOME I get.

tonycox commented:
@ramkrish86 what is your environment?

ramkrish86 commented:
It is CentOS Linux release 7.0.1406 (Core).

ramkrish86 commented:
> Results :

Tests run: 2, Failures: 0, Errors: 0, Skipped: 0

This is what I get as test result.

ramkrish86 commented:
Thanks to @tonycox  for helping me in reproducing this error. Changing to JDK 7 creates this issue and it creates due to permGen space running out of memory. I don't have a soln for this. It runs with JDK 8 with no hassles. Any inputs here?

tonycox commented:
@ramkrish86 Try to add 
```
-XX:+HeapDumpOnOutOfMemoryError 
-XX:HeapDumpPath="/tmp"
``` 
as parameter of JVM

fhueske commented:
Hi @ramkrish86, @tonycox could you reproduce the OOME on a local setup? 
We might want to compare the new HBase InputFormat and the existing `TableInputFormat` which does not fail in tests.

ramkrish86 commented:
I tried to compare TableInputFormat and the new one. But the interesting part is HBaseInputformat does not fail in jdk 8 which was my default in my test env.

ramkrish86 commented:
I tried multiple ways to take the dump file from this mvn test command run. I get a hprof file which on opening in heap dump analyser throws EOF exception or NPE exception. 
@tonycox - were you able to get any heap dump?

tonycox commented:
@ramkrish86 You can use jhat to read hprof dump file. 
If set `-XX:MaxPermSize=128m`
```xml
<plugin>
	<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-surefire-plugin</artifactId>
	<version>2.19.1</version>
	<configuration>
		<argLine>-XX:MaxPermSize=128m</argLine>
		<!-- Enforce single fork execution due to heavy mini cluster use in the tests -->
		<forkCount>1</forkCount>
	</configuration>
</plugin>
```
all tests will pass

ramkrish86 commented:
Even jhat was not able to view the file as it had a problem in parsing the hprof file. So my question is if MaxPermSize is 128M - why does it work with jdk 8?

ramkrish86 commented:
@fhueske  - so are you ok with @tonycox  suggestion of setting MAxPermSize for hbase module?

tonycox commented:
@ramkrish86 java8 has different memory model, https://dzone.com/articles/java-8-permgen-metaspace

ramkrish86 commented:
I think that is not the only reason. Some where either these tests are creating more static objects or bigger objects that live for the life time of the JVM. May be this test exposes it. Actually this change in maxPermsize calls for a discussion on how the 'mvn' settings should be and what is the minimum size of heap required. Also in other projects the default jdk version is now moved to 8 atleast for the trunk version. I think similar thing can be done here. Thanks @tonycox .

ramkrish86 commented:
@fhueske - Are you fine with that pom change? If so we can get this in.

fhueske commented:
It's not only Java 7. For Hadoop 2.4.1 and Java 7, the builds pass. So it seems to be a combination of both parameters. Anyway, I think we can configure the JVM for the tests. 

It would be good to check if @rmetzger has any objections. 
He knows Maven and testing infrastructure much better than I do.

rmetzger commented:
I think its fine to change the JVM settings for the hbase module. Other modules such as elasticsearch are also doing that.

fhueske commented:
Thanks Robert! 
I'll let the test run again and will merge if everything passes. Thanks @tonycox and @ramkrish86 for digging into this.

ramkrish86 commented:
Thanks to all @fhueske , @tonycox  and @wuchong for helping in getting this in.
